### PR TITLE
Fault and Fault_Drawer (3 MATCHING and 1 NON_MATCHING)

### PR DESCRIPTION
**Pull request by hensldm**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

Matched the last two functions in fault_drawer as well as one in fault (however that one requires in function bss, so going to keep it NON_MATCHING for now). Also got the last function in fault very close, but still having regalloc issues (scratch: https://decomp.me/scratch/x8y6X).

### CODE REVIEW

File: src/boot_O2_g3/fault_drawer.c [hunk 1]
```diff
@@ -142,12 +172,57 @@ void FaultDrawer_FillScreen() {
...
+                    FaultDrawer_FillScreen();
+                }
+                sFaultDrawContext->cursorY = sFaultDrawContext->yStart;
+            }
+        }
+    }
+
+    osWritebackDCacheAll();
+
+    return arg;
```
Comment (isghj5) [inline src/boot_O2_g3/fault_drawer.c:219]: ... It returns its own un-modified parameter?

File: src/boot_O2_g3/fault_drawer.c [hunk 2]
```diff
@@ -16,7 +18,7 @@ FaultDrawer sFaultDrawerDefault = {
     GPACK_RGBA5551(0, 0, 0, 0),         // backColor
     22,                                 // cursorX
     16,                                 // cursorY
-    (u32*)sFaultDrawerFont,             // font
+    sFaultDrawerFont,                   // font
     8,                                  // charW
     8,                                  // charH
     0,                                  // charWPad
```
Comment (EllipticEllipsis) [inline src/boot_O2_g3/fault_drawer.c:24]: Could you label the colours too? Most likely they're the same as the [ASCII terminal ones](https://en.wikipedia.org/wiki/ANSI_escape_code#Colors), i.e.
Black
Red
Green
Yellow
Blue
Magenta
Cyan
White
and two greys.

**Review by EllipticEllipsis — COMMENTED** [review]
Nice work, just some minor suggestions.

File: src/boot_O2_g3/fault.c [hunk 3]
```diff
@@ -751,8 +750,10 @@ void Fault_ProcessClients(void) {
 }
 
 #ifdef NON_MATCHING
-// regalloc and ordering differences around the two bool variables (faultCustomOptions and faultCopyToLog)
+// needs in-function static bss
 void Fault_SetOptionsFromController3(void) {
+    static u32 faultCustomOptions;
+
     Input* input3;
```
Comment (Thar0) [inline src/boot_O2_g3/fault.c:757]: We typically don't separate `static` and local variables
```suggestion
    static u32 faultCustomOptions;
    Input* input3;
```

File: src/boot_O2_g3/fault.c [hunk 4]
```diff
@@ -762,21 +763,23 @@ void Fault_SetOptionsFromController3(void) {
...
 
     if (faultCustomOptions) {
         graphPC = sGraphThread.context.pc;
         graphRA = sGraphThread.context.ra;
         graphSP = sGraphThread.context.sp;
-        if (CHECK_BTN_ALL(input3->press.button, BTN_R)) {
+        if (CHECK_BTN_ALL(input3->cur.button, BTN_R)) {
+            static u32 faultCopyToLog;
+
             faultCopyToLog = !faultCopyToLog;
```
Comment (Thar0) [inline src/boot_O2_g3/fault.c:776]: imo it looks cleaner this way
```suggestion
            static u32 faultCopyToLog = !faultCopyToLog;
```

File: src/boot_O2_g3/fault_drawer.c [hunk 5]
```diff
@@ -74,7 +77,36 @@ void FaultDrawer_DrawRecImpl(s32 xStart, s32 yStart, s32 xEnd, s32 yEnd, u16 col
...
+    const u32* dataPtr = &sFaultDrawContext->font[(((c / 8) * 16) + ((c & 4) >> 2))];
+    u16* fb = sFaultDrawContext->fb + (sFaultDrawContext->w * cursorY) + cursorX;
+
+    if ((sFaultDrawContext->xStart <= cursorX) &&
+        ((sFaultDrawContext->charW + cursorX - 1) <= sFaultDrawContext->xEnd) &&
+        (sFaultDrawContext->yStart <= cursorY) &&
+        ((sFaultDrawContext->charH + cursorY - 1) <= sFaultDrawContext->yEnd)) {
+        for (y = 0; y < sFaultDrawContext->charH; y++) {
+            u32 mask = 0x10000000 << shift;
+            data = *dataPtr;
```
Comment (Thar0) [inline src/boot_O2_g3/fault_drawer.c:96]:
```suggestion
            u32 mask = 0x10000000 << shift;

            data = *dataPtr;
```

**Review by Thar0 — CHANGES_REQUESTED** [review]
Little things

File: src/boot_O2_g3/fault.c (hunk 4 above)
Comment (hensldm) [reply src/boot_O2_g3/fault.c:776]: Apparently this is not allowed.
image

Comment (Thar0) [reply src/boot_O2_g3/fault.c:775]: Oh right, I misread what was going on here, sorry. Now that I read it properly I think it's best left as-is.

**Review by AngheloAlf — APPROVED** [review]

**Comment by hensldm** [comment]
Was able to match `Fault_FindNextStackCall`, so the only thing left in fault is the NON_MATCHING which matches but needs in function static bss.

**Review by engineer124 — APPROVED** [review]
lgtm

**Review by EllipticEllipsis — APPROVED** [review]
I'd have probably written the greys as light and dark, but I'm not particularly bothered since it's obvious from looking at the colour that they're different.

**Comment by hensldm** [comment]
Oh that is a good idea to distinguish the grays. Just added that real quick.

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**